NitroxClient/MonoBehaviours/MovementReplicator.cs

```diff
@@ -0,0 +1,212 @@
+                }
+                latestLatencyBumpTime = currentTime;
+                maxLatencyDetectedRecently = 0;
+            }
+        }
+
+        float occurrenceTime = time + INTERPOLATION_TIME + maxAllowedLatency;
+
+        // Cleaning any previous value change that would occur later than the newly received snapshot
+        while (buffer.Last != null && buffer.Last.Value.IsOlderThan(occurrenceTime))
+        {
+            buffer.RemoveLast();
+        }
+
+        buffer.AddLast(new Snapshot(movementData, occurrenceTime));
+    }
+
+    public void ClearBuffer() => buffer.Clear();
+
+    public void Start()
+    {
+        if (!gameObject.TryGetNitroxId(out objectId))
+        {
+            Log.Error($"Can't start a {nameof(MovementReplicator)} on {name} because it doesn't have an attached: {nameof(NitroxEntity)}");
+            return;
+        }
+
+        rigidbody = GetComponent<Rigidbody>();
+        if (gameObject.TryGetComponent(out NitroxCyclops nitroxCyclops))
+        {
+            nitroxCyclops.SetReceiving();
+        }
+        else
+        {
+            if (gameObject.TryGetComponent(out WorldForces worldForces))
+            {
+                worldForces.enabled = false;
+            }
+            rigidbody.isKinematic = false;
+        }
+
+        MovementBroadcaster.RegisterReplicator(this);
+    }
+
+    public void OnDestroy()
+    {
+        if (gameObject.TryGetComponent(out NitroxCyclops nitroxCyclops))
+        {
+            nitroxCyclops.SetBroadcasting();
+        }
+        else
+        {
+            if (gameObject.TryGetComponent(out WorldForces worldForces))
+            {
+                worldForces.enabled = true;
+            }
+        }
+
+        MovementBroadcaster.UnregisterReplicator(this);
+    }
+
+    public void Update()
+    {
+        if (buffer.Count == 0)
+        {
+            return;
+        }
+
+        float currentTime = CurrentTime;
+
+        // Sorting out expired nodes
+        while (buffer.First != null && buffer.First.Value.IsExpired(currentTime))
+        {
+            buffer.RemoveFirst();
+        }
+
+        LinkedListNode<Snapshot> firstNode = buffer.First;
+        if (firstNode == null)
+        {
+            return;
+        }
+
+        // Current node is not useable yet
+        if (firstNode.Value.IsOlderThan(currentTime))
+        {
+            return;
+        }
+
+        // Purging the next nodes if they should have already happened (we still have an expiration margin for the first node so it's fine)
+        while (firstNode.Next != null && !firstNode.Next.Value.IsOlderThan(currentTime))
+        {
+            firstNode = firstNode.Next;
+            buffer.RemoveFirst();
+        }
+
+        LinkedListNode<Snapshot> nextNode = firstNode.Next;
+
+        // Current node is fine but there's no next node (waiting for it without dropping current)
+        if (nextNode == null)
+        {
+            return;
```

Comment:
This would happen if the simulated player:
- stopped sending packets (e.g. game freezing)
- disconnected
or if the local player is experiencing an unplanned latency

In those cases, we either don't want extrapolation at all because it's generally too risky to move the vehicle where it probably isn't (extrapolating for 1 second at a speed of 10m/s means 10 meters error, and with this new system, it'd possibly end up with the cyclops going through a wall, while the local player is still inside of the cyclops) or the movement correction would most likely look even worse than the simple stutter that would happen if there's an unplanned latency.
For bad connections, they have two settings to adjust to enjoy a smooth experience.